notebooks/02-ingestion/10-updating-data.ipynb

```diff
@@ -0,0 +1,1188 @@
+    "- The flight was incorrectly recorded as using the plane with Tail_Number 'N063AA' \n",
+    "- It actually left late because a change of plane to Tail_Number 'N073AA' \n",
+    "- The key to identify a single row in this table involves, __time (departure time), airline and flight_number\n",
+    "\n",
+    "In a traditional SQL database you would do something like:\n",
+    "```\n",
+    "UPDATE \"example-flights-updates\"\n",
+    "    SET \"tail_number\"='N073AA'\n",
+    "  WHERE \"__time\"= '2005-11-01T00:15:00.000Z' \n",
+    "        AND \"airline\"='AA' \n",
+    "        AND \"flight_number\"=513\n",
+    "```\n",
+    "\n",
+    "Run the following cell to look at that row as it exists in the table before an update:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "c0d5f243-63f2-4edb-9a29-2c80ec42f723",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "\n",
+    "sql='''\n",
+    "SELECT * FROM \"example-flights-updates\" \n",
+    "WHERE \"__time\"='2005-11-01T00:15:00.000Z' \n",
+    "  AND \"airline\"='AA' \n",
+    "  AND \"flight_number\"=513\n",
+    "'''\n",
+    "display.sql(sql)\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "a13e5c40-89d1-4f3d-b8f9-f5e3e1f33a3e",
+   "metadata": {},
+   "source": [
+    "### Druid segment overlay logic\n",
+    "At query time Druid selects the segments that are relevant to a query. Typically this involves a set of segment files that cover regular time intervals based on the segment granularity that was ingested. Up to this point you have ingested the data into DAY partitions. \n",
+    "\n",
+    "One of the lesser known capabilities in Druid is the ability to overlay segments of different granularity. This capability allows you to change the contents of a portion of the timeline within a particular segment partition by adding a smaller time granularity segment on top of existing data. "
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "26b2aff3-b485-42ac-868a-a75ce893deb2",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# The smallest slice of a time chunk that can be updated is 1 second\n",
+    "# Druid's segment overlapping and versioning strategy requires that the new segment include\n",
+    "# all of the rows in the time interval where a row is going to be changed. \n",
+    "# This SQL shows what other rows occur in the same second? \n",
+    "sql='''\n",
+    "SELECT * FROM \"example-flights-updates\" \n",
+    "WHERE TIME_IN_INTERVAL(\"__time\", '2005-11-01T00:15:00.000Z/PT1S') \n",
+    "'''\n",
+    "display.sql(sql)\n"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "29c99123-3092-4102-b3f4-20bb804227cb",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "#before updating anything, get a count and a checksum so we can validated data consistency\n",
```

Comment:
There aren't that many steps. It is just going through to show that since we are replacing a second of the timeline we need to include anything else within that second, so it shows which other rows are in that second. It also shows that it is a true update, by showing that the totals for the day did not change, not even the row count. That's why I'm doing the aggregation as well.
The actual UPDATE follows this with a full explanation of each part of the REPLACE statement that is needed. This is really the only step needed for the update.